### PR TITLE
Add an option for kedro new to skip telemetry

### PR DESCRIPTION
**Pull request by lrcouto**
## Description

Adds a new `--telemetry` or `-tc` flag for `kedro new` which allows the user to sign their consent for usage analytics collection on the moment of the creation of the project, and skipping the prompt for analytics collection on the moment where the `kedro` command is first invoked inside the project directory.

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [x] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [x] Added tests to cover my changes
- [x] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

File: kedro/framework/cli/starters.py [hunk 1]
```diff
@@ -269,6 +275,7 @@ def new(  # noqa: PLR0913
     checkout: str,
     directory: str,
     example_pipeline: str,  # This will be True or False
```
Comment (DimedS) [inline kedro/framework/cli/starters.py:277]:
```suggestion
    example_pipeline: str,
    telemetry_consent: str,
```
It seems the comment is outdated. Previously, the `example_pipeline` was returned with Click preprocessed to contain True/False values. However, it now retains its original 'y'/'no' values. It would be best to remove this comment.

File: kedro/framework/cli/starters.py [hunk 2]
```diff
@@ -876,7 +890,9 @@ def _validate_range(start: Any, end: Any) -> None:
     return selected
 
 
-def _create_project(template_path: str, cookiecutter_args: dict[str, Any]) -> None:
+def _create_project(
+    template_path: str, cookiecutter_args: dict[str, Any], telemetry_consent: str
```
Comment (DimedS) [inline kedro/framework/cli/starters.py:894]: Is it `telemetry_consent: str|None`?

**Review by DimedS — CHANGES_REQUESTED** [review]
Thank you for the PR, @lrcouto! Looks good to me, I've left some minor comments.

**Review by noklam — COMMENTED** [review]
Not much comments about the implementation since it's quite similar to example pipelines, but more on why we want to keep `yes` `no` instead of just a default boolean, similar to `kedro run --async`

File: docs/source/get_started/new_project.md [hunk 3]
```diff
@@ -112,6 +112,11 @@ You can also enter this in a single line as follows:
 kedro new --name=testproject --tools=lint,docs,pyspark --example=n
 ```
 
+### Telemetry consent
+
+The `--telemetry` flag offers the option to register consent to have user analytics collected in the moment of the creation of the project. This option bypasses the prompt to collect analytics that would otherwise appear on the moment the `kedro` command is invoked for the first time inside the project. In case the `--telemetry` flag is not used, the user will be prompted to accept or reject analytics collection as usual.
+
+When creating your new Kedro project, use the values `yes` or `no` to register consent to have user analytics collected for this specific project.
```
[the remark this answers is not in the document]
Comment (DimedS) [reply docs/source/get_started/new_project.md:119]: I believe here and for example_pipeline, we should leave it as it is: with yes/no option. Because if you want to choose `--telemetry=yes`,` --skip-telemetry` will not provide you that option.

Comment (merelcht) [inline docs/source/get_started/new_project.md:119]: Maybe adding this sentence clarifies the behaviour? 

```suggestion
When creating your new Kedro project, use the values `yes` or `no` to register consent to have user analytics collected for this specific project. Selecting `yes` means you consent to your data being collected, whereas `no` means you do not consent and no data will be collected. 
```

File: tests/framework/cli/test_starters.py [hunk 4]
```diff
@@ -1562,3 +1562,119 @@ def test_convert_tool_short_names_to_numbers_with_duplicates(self):
         selected_tools = "lint,test,tests"
         result = _convert_tool_short_names_to_numbers(selected_tools)
         assert result == ["1", "2"]
+
+
+@pytest.mark.usefixtures("chdir_to_tmp")
```
[the remark this answers is not in the document]
Comment (lrcouto) [reply tests/framework/cli/test_starters.py:1576]: Moving the tests related to tools to their own file and maybe moving auxiliary functions to their own file as well would probably solve half ot it.

**Review by astrojuanlu — CHANGES_REQUESTED** [review]
I was discussing this with @iamelijahko and I am not 100 % sure we are solving the root cause here. Sorry for being so late to the party, should have anticipated this earlier.

The original problem is that Kedro gets stuck in certain non-interactive environments #1640. The current workarounds are 

1. `echo "consent: false" > .telemetry` and then continue. Unintuitive.
2. Uninstall `kedro-telemetry` altogether. We've been counseling this to Databricks users.

Discussion about this issue is a bit scattered, there's a bit more context in https://github.com/kedro-org/kedro/issues/2867 of course.

Adding this flag to `kedro new` (whatever name it takes, we can discuss about that later) would help users that run `kedro new` on these unattended environments, but my understanding is that the large majority actually run `kedro run` and the like. So this would not help them.

In other words: I don't recall seeing evidence that users _create_ new projects on CI environments.

Can we discuss this a bit further before proceeding?

File: kedro/framework/cli/starters.py [hunk 5]
```diff
@@ -67,6 +67,11 @@
 """
 EXAMPLE_ARG_HELP = "Enter y to enable, n to disable the example pipeline."
 
+TELEMETRY_ARG_HELP = """Allow or not allow Kedro to collect usage analytics.
+We cannot see nor store information contained into a Kedro project. Opt in with "yes"
+and out with "no".
+"""
```
[the remark this answers is not in the document]
Comment (datajoely) [reply kedro/framework/cli/starters.py:73]: I would copy the definition from the regex function as it's more descriptive:

`It must contain only y, n, YES, or NO (case insensitive).`

**Comment by datajoely** [comment]
@astrojuanlu I take your point - but I think we need this at creation time AND at `kedro <within project commands>` time 

I wonder if we can have a catch all workflow for any command that triggers the telemetry consent prompt to accept this flag 🤔

**Comment by astrojuanlu** [comment]
That's fair enough - if so, I contend that whatever we do

- We better do it with the whole CLI in mind,
- Probably as a PR to `kedro-telemetry` if at all possible?? Given that it's the one prompting for consent (otherwise this creates some weird separation of concerns issues), and
- Bearing in mind that at some point these flags might not make sense if we move towards non-interactive consent by default

**Comment by noklam** [comment]
https://github.com/kedro-org/kedro/issues/1640 was the original issue, I did recall the workaround of `yes | kedro new`. I agree this doesn't solve the whole problem particular for `kedro run`.

**Comment by merelcht** [comment]
I see the point of covering the full CLI commands, but my guess is that if we add a flag to `kedro new` and `kedro run` we solve about 90% of the cases where a kedro command is run in a CI type of environment. 

> Probably as a PR to kedro-telemetry if at all possible?? Given that it's the one prompting for consent (otherwise this creates some weird separation of concerns issues)

I like the sound of this, but have no idea if that's possible. 

My suggestion would be to merge this PR and add a similar flag to `kedro run`. We can do a spike to figure out if we could do it for the whole CLI from `kedro-telemetry`, but IMO that's lower priority.

**Comment by datajoely** [comment]
@merelcht I agree! 80/20 approach is fine.

File: docs/source/get_started/new_project.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (astrojuanlu) [reply docs/source/get_started/new_project.md]: I'm happy with `--telemetry yes/no` 👍🏽

**Review by astrojuanlu — APPROVED** [review]
Let's get this merged!
